BASE/Test/ServerTelemetryChannel.Test/Shared.Tests/Implementation/TelemetryBufferTest.cs

```diff
@@ -379,6 +380,137 @@ public void DoesNotContinueOnCapturedSynchronizationContextToImprovePerformance(
+                {
+                    OnSerializeAsync = (telemetry, cancellationToken) =>
+                    {
+                        serializedTelemetry = new List<ITelemetry>(telemetry);
+                        return Task.FromResult(true);
+                    }
+                };
+
+                var telemetryBuffer = new TelemetryBuffer(serializer, new StubApplicationLifecycle());
+
+                var expectedTelemetry = new StubTelemetry();
+                telemetryBuffer.Process(expectedTelemetry);
+
+                telemetryBuffer.ManualFlushAsync(new CancellationToken()).GetAwaiter().GetResult();
+
+                Assert.AreSame(expectedTelemetry, serializedTelemetry.Single());
+            }
+
+            [TestMethod]
+            [Timeout(10000)]
+            public async Task EmptiesBufferAfterSerialization()
+            {
+                var serializer = new StubTelemetrySerializer
+                {
+                    OnSerializeAsync = (telemetry, cancellationToken) =>
+                    {
+                        return Task.FromResult(true);
+                    }
+                };
+
+                var buffer = new TelemetryBuffer(serializer, new StubApplicationLifecycle());
+                buffer.Capacity = 10;
+                buffer.Process(new StubTelemetry());
+
+                await buffer.ManualFlushAsync(new CancellationToken());
+
+                Assert.AreEqual(0, buffer.Count());
+            }
+
+            [TestMethod]
+            [Timeout(10000)]
+            public async Task DoesNotContinueOnCapturedSynchronizationContextToImprovePerformance()
+            {
+                var serializer = new StubTelemetrySerializer
+                {
+                    OnSerializeAsync = (telemetry, cancellationToken) =>
+                    {
+                        return Task.FromResult(true);
+                    }
+                };
+
+                var buffer = new TelemetryBuffer(serializer, new StubApplicationLifecycle());
+                buffer.Process(new StubTelemetry());
+
+                bool postedBack = false;
+                using (var context = new StubSynchronizationContext())
+                {
+                    context.OnPost = (callback, state) =>
+                    {
+                        postedBack = true;
+                        callback(state);
+                    };
+
+                    await buffer.ManualFlushAsync(new CancellationToken());
+
+                    Assert.IsFalse(postedBack);
+                }
+            }
+
+            [TestMethod]
+            public void SetCancellationTokenOnManualFlushAsync()
+            {
+                var telemetryBuffer = new TelemetryBuffer(new StubTelemetrySerializer(), new StubApplicationLifecycle());
+                var task = telemetryBuffer.ManualFlushAsync(new CancellationToken(true));
+                Thread.Sleep(TimeSpan.FromMilliseconds(100));
```

Comment:
This is fragile (subject to timing variance).
Can you "await" the task and catch OperationCanceledException?